extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -0,0 +1,762 @@
+  {
+    PartitionResource resource = partitionResources.get(partition);
+    if (resource == null) {
+      throw new ISE("Partition [%s] has not been assigned", partition);
+    }
+
+    log.debug(
+        "Seeking partition [%s] to [%s]",
+        partition.getPartitionId(),
+        sequenceNumber != null ? sequenceNumber : iteratorEnum.toString()
+    );
+
+    resource.shardIterator = kinesis.getShardIterator(
+        partition.getStream(),
+        partition.getPartitionId(),
+        iteratorEnum.toString(),
+        sequenceNumber
+    ).getShardIterator();
+
+    checkPartitionsStarted = true;
+  }
+
+  private void filterBufferAndResetFetchRunnable(Set<StreamPartition<String>> partitions) throws InterruptedException
+  {
+    scheduledExec.shutdown();
+
+    try {
+      if (!scheduledExec.awaitTermination(EXCEPTION_RETRY_DELAY_MS, TimeUnit.MILLISECONDS)) {
+        scheduledExec.shutdownNow();
+      }
+    }
+    catch (InterruptedException e) {
+      log.warn(e, "InterruptedException while shutting down");
+      throw e;
+    }
+
+    scheduledExec = Executors.newScheduledThreadPool(
+        fetchThreads,
+        Execs.makeThreadFactory("KinesisRecordSupplier-Worker-%d")
+    );
+
+    // filter records in buffer and only retain ones whose partition was not seeked
+    BlockingQueue<OrderedPartitionableRecord<String, String>> newQ = new LinkedBlockingQueue<>(recordBufferSize);
+    records
+        .stream()
+        .filter(x -> !partitions.contains(x.getStreamPartition()))
+        .forEachOrdered(newQ::offer);
+
+    records = newQ;
+
+    // restart fetching threads
+    partitionResources.values().forEach(x -> x.started = false);
+    checkPartitionsStarted = true;
+  }
+
+  @Nullable
+  private String getSequenceNumberInternal(StreamPartition<String> partition, ShardIteratorType iteratorEnum)
+  {
+
+    String shardIterator = null;
+    try {
+      shardIterator = kinesis.getShardIterator(
+          partition.getStream(),
+          partition.getPartitionId(),
+          iteratorEnum.toString()
+      ).getShardIterator();
+    }
+    catch (ResourceNotFoundException e) {
+      log.warn(e, "Caught ResourceNotFoundException while getting shardIterator");
+    }
+
+    return getSequenceNumberInternal(partition, shardIterator);
+  }
+
+  @Nullable
+  private String getSequenceNumberInternal(StreamPartition<String> partition, String shardIterator)
+  {
+    long timeoutMillis = System.currentTimeMillis() + fetchSequenceNumberTimeout;
+
+    while (shardIterator != null && System.currentTimeMillis() < timeoutMillis) {
+
+      if (closed) {
+        log.info("KinesisRecordSupplier closed while fetching sequenceNumber");
+        return null;
+      }
+
+      GetRecordsResult recordsResult;
+      try {
+        recordsResult = kinesis.getRecords(new GetRecordsRequest().withShardIterator(shardIterator).withLimit(1000));
+      }
+      catch (ProvisionedThroughputExceededException e) {
```

Comment:
We should let the user know when we get this exception so that they understand what's happening (and maybe can take corrective action by scaling up).